agent/engine/common_test.go

```diff
@@ -0,0 +1,266 @@
+				{
+					Dir:  reflect.SelectRecv,
+					Chan: reflect.ValueOf(done),
+				},
+			})
+			if ndx == 1 {
+				break
+			}
+		}
+	}()
+	return func() {
+		done <- true
+	}
+}
+
+func verifyTaskIsRunning(stateChangeEvents <-chan statechange.Event, testTasks ...*api.Task) error {
+	for {
+		select {
+		case event := <-stateChangeEvents:
+			if event.GetEventType() == statechange.TaskEvent {
+				taskEvent := event.(api.TaskStateChange)
+				for i, task := range testTasks {
+					if taskEvent.TaskARN != task.Arn {
+						continue
+					}
+					if taskEvent.Status == api.TaskRunning {
+						if len(testTasks) == 1 {
+							return nil
+						}
+						testTasks = append(testTasks[:i], testTasks[i+1:]...)
+					} else if taskEvent.Status > api.TaskRunning {
+						return fmt.Errorf("Task went straight to %s without running, task: %s", taskEvent.Status.String(), task.Arn)
+					}
+				}
+			}
+		}
+	}
+}
+
+func verifyTaskIsStopped(stateChangeEvents <-chan statechange.Event, testTasks ...*api.Task) {
+	for {
+		select {
+		case event := <-stateChangeEvents:
+			if event.GetEventType() == statechange.TaskEvent {
+				taskEvent := event.(api.TaskStateChange)
+				for i, task := range testTasks {
+					if taskEvent.TaskARN == task.Arn && taskEvent.Status >= api.TaskStopped {
+						if len(testTasks) == 1 {
+							return
+						}
+						testTasks = append(testTasks[:i], testTasks[i+1:]...)
+					}
+				}
+			}
+		}
+	}
+}
+
+// waitForTaskStoppedByCheckStatus verify the task is in stopped status by checking the KnownStatusUnsafe field of the task
+func waitForTaskStoppedByCheckStatus(task *api.Task) {
+	for {
+		if task.GetKnownStatus() == api.TaskStopped {
+			return
+		}
+		time.Sleep(50 * time.Millisecond)
+	}
+}
+
+// validateContainerRunWorkflow validates the container create and start workflow
+// for a test task without any resources (such as ENIs).
+//
+// The createdContainerName channel is used to emit the container name from the
+// create operation. It can be used to validate that the name of the container
+// removed matches with the generated container name during cleanup operation in the
+// test.
+func validateContainerRunWorkflow(t *testing.T,
+	container *api.Container,
+	task *api.Task,
+	imageManager *MockImageManager,
+	client *MockDockerClient,
+	roleCredentials *credentials.TaskIAMRoleCredentials,
+	createStartEventsReported sync.WaitGroup,
+	eventStream chan DockerContainerChangeEvent,
+	createdContainerName chan<- string,
+	assertions func(),
+) {
+	imageManager.EXPECT().AddAllImageStates(gomock.Any()).AnyTimes()
+	client.EXPECT().PullImage(container.Image, nil).Return(DockerContainerMetadata{})
+	imageManager.EXPECT().RecordContainerReference(container).Return(nil)
+	imageManager.EXPECT().GetImageStateFromImageName(gomock.Any()).Return(nil)
+	client.EXPECT().APIVersion().Return(defaultDockerClientAPIVersion, nil)
+	dockerConfig, err := task.DockerConfig(container, defaultDockerClientAPIVersion)
+	if err != nil {
+		t.Fatal(err)
```

Comment:
Can you change this to use `require`?

Reply:
We're explicitly not doing `require` here. There's some weirdness in the return type, which doesn't treat `nil` value for this type as `nil`.